### PR TITLE
Implement SDL 0090 SDLHapticHitTester Protocol and Implement SDL 0081 Automatic Haptic Interface Detection

### LINKED ISSUE #698: [SDL 0081] Automatic Haptic Interface Detection for UIKit based VPM / Nav applications
Proposal: [SDLInterfaceManager](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0081-SDLInterfaceManager.md)

> The SendHapticData RPC will allow video projection apps to tell SDLCore where their controls are located. This allows VPM apps to integrate with the OEM haptic engine in a way similar to template SDL apps. This proposal aims to implement the necessary supporting proxy logic so that applications can easily utilize the OEM haptic feedback mechanism.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/236

Steering Committee Decision:

> The Steering Committee has agreed to accept this proposal with revisions, requesting that the manager use a protocol-oriented design pattern. The Steering Committee will also have an additional workshop to discuss a larger video projection mode feature.
These changes have now been reflected in .md file.

### LINKED ISSUE #728: [SDL 0090] SDLHapticHitTester Protocol
Proposal: [SDLHapticHitTester Protocol](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0090-SDLHapticHitTesterProtocol.md)

> SDL-0081 introduced the SDL Haptic Interface Manager. Since this manager stores the location of all haptic UI components, it can be used to ease integration with apps that utilize it when responding to touch events. This proposal aims to introduce a protocol that takes a SDLTouch object, and returns the selected haptic view when applicable.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/270

Steering Committee Decision:

> The Steering Committee has voted to accept this proposal with the revisions included in Joel Fischer’s comment.  It was noted that the functions included in the changes to the `SDLTouchManagerDelegate` already exist, so only the new parameter would need to be added.  Since this feature is planned to be included in the iOS 5.0 release, breaking changes could be included.

>The proposal .md was updated on September 8, 2017 to include the agreed upon revisions.

### PR DESCRIPTION
**Pull request by ChrisB-Elektrobit**
Fixes #698 
Fixes #728 

This PR is **[ready]** for review.

### Notes
Alternative implementations for the `viewForSDLTouch()` function were considered.
The current approach was considered more readable. Pending code review feedback.

### Risk
This PR makes **[minor]** API changes.

### Testing Plan
Testing was mainly done through a modified Example app that is already provided as a part of the sdl_ios repository.  From that test application, an instance of `HapticManager` was created with a UI view window provided as a parameter.  We observed that the `HapticManager` was able to successfully extract the geometry of focusable elements and use these while triggering a call to SendHapticData RPC.  In our test setup, we were able to see the SendHapticData RPC being received by a simulated head unit with the proper parameters as extracted by the `HapticManager`.  In addition, the `HapticManager`'s `viewForSDLTouch()` function was fed with a number of (x,y) coordinates to test the second use case as described above.  We tested points both inside and outside of the various rectangles in order to confirm that the appropriate UI views were either selected or remained unselected.

### Summary
The changes included in this pull request are intended to provide a convenience layer for mobile apps running on iOS that would like to use haptic input via the SendHapticData RPC. The main use case provided the capability to scan a provided view for focusable elements and extract the view's geometry.
 The extracted geometry is then passed on to the SendHapticData RPC. Additionally, a use case is supported where based on provided x,y coordinates, a search is performed for a UI element that would contain the provided x,y coordinates within its geometry. Both use cases are implemented as part of a new class called the `HapticManager`.

##### Breaking Changes
* No Breaking Changes

##### Enhancements
* A new class `HapticManager` was implemented. Because of ongoing discussions as to where to call the functionality from, no further integration was done. The decision on the final integration point will be up to the project maintainers.
* SDL-0081 introduces the SDL Haptic Interface Manager. This manager provides functionality to calculate and store the locations of all haptic UI components. It can be used to ease integration with apps that utilize it when responding to touch events.
* SDL-0090 introduces SDL haptic hit tester functionality.  This includes in interface into the Haptic Manager that takes a SDLTouch object, and returns the selected haptic view when applicable.

##### Bug Fixes
* No bug fixes

### Tasks Remaining:
- [ ] Add unit tests

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by joeljfischer** [comment]
@ChrisB-Elektrobit This PR has been marked for review, but the unit tests are necessary before review can begin. Please provide comprehensive unit tests as soon as possible so that we can provide a thorough review of this PR.

**Comment by joeljfischer** [comment]
@ChrisB-Elektrobit Additionally, even though the example app used to test should not be a part of the repository, could you please provide a zip in a comment here containing the project to aid in our tests?

**Comment by ChrisB-Elektrobit** [comment]
@joeljfischer , we added some unit tests for the haptic manager and merged the base branch "release/5.0.0" back into our feature branch.  I am also attaching the example app we used for testing as requested.  You can replace your “SmartDeviceLink_Example” directory with the contents of the zipfile and build it if you want to try it out.
[SmartDeviceLink_Example.zip](https://github.com/smartdevicelink/sdl_ios/files/1329941/SmartDeviceLink_Example.zip)

**Comment by joeljfischer** [comment]
@ChrisB-Elektrobit 👍 Great, I will start reviewing and testing ASAP

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    beforeEach(^{
+        uiWindow = [[UIWindow alloc] init];
+        uiViewController = [UIViewController alloc];
+        
+        [uiWindow addSubview:uiViewController.view];
+        
+        OCMExpect( [[sdlManagerMock stub] sendRequest:[OCMArg checkWithBlock:^BOOL(id value){
+            BOOL isFirstArg = [value isKindOfClass:[SDLSendHapticData class]];
+            if(isFirstArg)
+            {
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: Stick to braces on the same line as the conditional

File: SmartDeviceLink/SDLHapticInterface.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//
+//  Copyright © 2017 smartdevicelink. All rights reserved.
+//
+
+#import <UIKit/UIKit.h>
+#import "SDLManager.h"
+
+@protocol SDLHapticInterface <NSObject>
+
+- (instancetype)initWithWindow:(UIWindow *)window  sdlManager:(SDLManager *)sdlManager;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticInterface.h]: I'm not 100% sure if this will take an `SDLManager` but at least the method should be `initWithWindow:connectionManager:` to match the other managers such as `SDLFileManager`.

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+SDLTouch* generateTouchEvent(int xCoord, int yCoord)
+{
+    unsigned long timeStamp = 10;
+    SDLTouchCoord* firstCoord = [[SDLTouchCoord alloc] init];
+    firstCoord.x = [NSNumber numberWithInt:xCoord];
+    firstCoord.y = [NSNumber numberWithInt:yCoord];
+    
+    SDLTouchEvent* firstTouchEvent = [[SDLTouchEvent alloc] init];
+    firstTouchEvent.touchEventId = @0;
+    firstTouchEvent.coord = [NSMutableArray arrayWithObject:firstCoord];
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: These can be ordinary `NSArray`, no need for the mutability.

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+#import "SDLManager.h"
+#import "SDLTouchCoord.h"
+#import "SDLTouchEvent.h"
+#import "SDLTouch.h"
+
+SDLTouch* generateTouchEvent(int xCoord, int yCoord)
+{
+    unsigned long timeStamp = 10;
+    SDLTouchCoord* firstCoord = [[SDLTouchCoord alloc] init];
+    firstCoord.x = [NSNumber numberWithInt:xCoord];
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: Use Boxing here `firstCoord.x = @(xCoord);`

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            if(isFirstArg)
+            {
+                capturedHapticRPC = value;
+            }
+            return YES;
+        }] withResponseHandler:[OCMArg any]]);
+
+    });
+    
+    context(@"when initialized with no view", ^{
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: Should this be "with no focusable views"?

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+QuickSpecBegin(SDLHapticManagerSpec)
+
+describe(@"the haptic manager", ^{
+    
+    __block UIWindow *uiWindow;
+    __block UIViewController *uiViewController;
+    
+    __block SDLHapticManager *hapticManager;
+    __block SDLSendHapticData* capturedHapticRPC;
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: rename `sentHapticRequest`

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    });
+    
+    context(@"when initialized with no view", ^{
+        beforeEach(^{
+            hapticManager = [[SDLHapticManager alloc] initWithWindow:uiWindow sdlManager:sdlManagerMock];
+        });
+        
+        it(@"should have no view", ^{
+            OCMVerify(sdlManagerMock);
+            expect([[capturedHapticRPC hapticRectData] count]).to(equal(0));
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: dot syntax `capturedHapticRPC.hapticRectData.count`

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        });
+    });
+    
+    context(@"when initialized with single view", ^{
+        beforeEach(^{
+            UITextField *textField1 = [[UITextField alloc]  initWithFrame:CGRectMake(101, 101, 50, 50)];
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: There should also be a test to make sure that the captured coordinates are correct.

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    
+    __block UIWindow *uiWindow;
+    __block UIViewController *uiViewController;
+    
+    __block SDLHapticManager *hapticManager;
+    __block SDLSendHapticData* capturedHapticRPC;
+    
+    __block id sdlManagerMock = OCMClassMock([SDLManager class]);
+    
+    beforeEach(^{
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: Reset the `hapticManager` to nil if it's reset in every context. Reset the `capturedHapticRPC` to nil here as well.

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    });
+        
+    context(@"when initialized with no views and then updated with two additional views", ^{
+        beforeEach(^{
+            hapticManager = [[SDLHapticManager alloc] initWithWindow:uiWindow sdlManager:sdlManagerMock];
+            
+            UITextField *textField1 = [[UITextField alloc]  initWithFrame:CGRectMake(101, 101, 50, 50)];
+            [uiViewController.view addSubview:textField1];
+            
+            UITextField *textField2 = [[UITextField alloc]  initWithFrame:CGRectMake(201, 201, 50, 50)];
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: Needs tests for both of these to have correct coordinates.

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        
+        it(@"should have two views", ^{
+            OCMVerify(sdlManagerMock);
+            expect([[capturedHapticRPC hapticRectData] count]).to(equal(2));
+        });
+     });
+    
+    context(@"when initialized with two views and then updated with one view removed", ^{
+        beforeEach(^{
+            UITextField *textField1 = [[UITextField alloc]  initWithFrame:CGRectMake(101, 101, 50, 50)];
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: See above on checking location

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+            [uiViewController.view addSubview:textField1];
+            
+            UITextField *textField2 = [[UITextField alloc]  initWithFrame:CGRectMake(201, 201, 50, 50)];
+            [uiViewController.view addSubview:textField2];
+            
+            hapticManager = [[SDLHapticManager alloc] initWithWindow:uiWindow sdlManager:sdlManagerMock];
+        });
+        
+        it(@"should return a view object", ^{
+
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: Bonus line

File: SmartDeviceLink/SDLHapticManager.m [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        
+    }
+    return self;
+}
+
+- (void)updateInterfaceLayout {
+    self.focusableViews = [NSMutableArray new];
+    [self parseViewHierarchy:[[self.projectionWindow subviews] lastObject]];
+    
+    NSUInteger preferredViewIndex = [self.focusableViews indexOfObject:[[[self.projectionWindow subviews] lastObject] preferredFocusedView]];
```
[the remark this answers is not in the document]
Comment (ChrisB-Elektrobit) [reply SmartDeviceLink/SDLHapticManager.m]: Correct, this will not support a `UIButton` if that button type does not support 'focusable' testing via the `canBecomeFocused` property.
In order to use a `UIButton` with the haptic manager, it will somehow need to be flagged as `canBecomeFocused`.
As long as a `UIButton` or any other special view is flagged properly, the `parseViewHierarchy` function should work as expected.
It seems like it might be cleaner to handle the special case logic for flagging a button outside of the haptic manager itself.

**Comment by ChrisB-Elektrobit** [comment]
> It appears that this proposal is intended for this release as well, but is not fully implemented in this PR. Specifically the touch manager changes are missing. Is this something that Livio is expected to write and test?

The touch manager changes are one of the integration points that we felt would be better left to Livio.  In order to implement the changes ourselves, we would need to understand a few more details for how the haptic manager should be integrated into the system.  The main point of interest here would be to understand where the haptic manager is created and to be able to get a reference to it in order to call `viewForSDLTouch()` from within the touch manager.  We would still be willing to write the code for this as well, but this probably could not be completed by Wednesday.

**Comment by joeljfischer** [comment]
Regarding the `UIButton` integration with haptic, that is needed here. I believe Xevo's proposal (by @davidswi) was to use a category on `UIButton`, but I can no longer find the post. If not, some swizzling may be needed, even though I generally very much attempt to avoid that. 

I'm okay with taking on the touch manager integration, but it means we'll need as much time as possible to integrate and test by Friday. Could the `UIButton` fix be figured out by EOD Wednesday? I'm available on Slack for further conversation as well.

File: SmartDeviceLink/SDLHapticManager.m [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    NSArray *focusableSubviews = [currentView.subviews filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(UIView *  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
+        return evaluatedObject.canBecomeFocused;
+    }]];
+    
+    //if current view is focusable and it doesn't have any focusable sub views then add the cuurent view and return
+    if (currentView.canBecomeFocused && focusableSubviews.count == 0) {
+        [self.focusableViews addObject:currentView];
+        return;
+    }
+    // if current view has focusable sub views parse them recursively
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticManager.m]: So only the focusable subviews get added? The top level focusable view does not? Is that how we're expecting that to work? There should be a test for this code as well.

File: SmartDeviceLink/SDLHapticManager.m [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    UIView *selectedView = nil;
+    
+    for (UIView *view in self.focusableViews) {
+        //Convert the absolute location to local location and check if that falls within view boundary
+        CGPoint localPoint = [view convertPoint:touch.location fromView:self.projectionWindow];
+        if ([view pointInside:localPoint withEvent:nil]) {
+            if (selectedView != nil) {
+                selectedView = nil;
+                break;
+                //the point has been indentified in two views. We cannot identify which with confidence.
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticManager.m]: Returning neither view is what we want here? Is that how `hitTest` works? Is this even possible considering the code in `sdl_parseViewHierarchy` will check subviews? There should be a unit test for this as well.

File: SmartDeviceLink/SDLHapticManager.m [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+            } else {
+                selectedView = view;
+            }
+        }
+    }
+    
+    if(selectedView != nil) {
+        NSUInteger selectedViewIndex = [self.focusableViews indexOfObject:selectedView];
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticManager.m]: Please use subscripting

File: SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+    
+    SDLTouchEvent* firstTouchEvent = [[SDLTouchEvent alloc] init];
+    firstTouchEvent.touchEventId = @0;
+    firstTouchEvent.coord = [NSArray arrayWithObject:firstCoord];
+    firstTouchEvent.timeStamp = [NSMutableArray arrayWithObject:@(timeStamp)];
+    SDLTouch* firstTouch = [[SDLTouch alloc] initWithTouchEvent:firstTouchEvent];
+    return firstTouch;
+}
+
+bool compareRectangle(SDLRectangle *sdlRectangle, CGRect cgRect)
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m]: This should probably return `BOOL` which is the obj-c boolean type.

File: SmartDeviceLink/SDLHapticManager.m [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+        SDLRectangle* rect = [[SDLRectangle alloc] initWithCGRect:(convertedRect)];
+        // using the view index as the id field in SendHapticData request (should be guaranteed unique)
+        NSUInteger rectId = [self.focusableViews indexOfObject:view];
+        SDLHapticRect *hapticRect = [[SDLHapticRect alloc] initWithId:(UInt32)rectId rect:rect];
+        [hapticRects addObject:hapticRect];
+    }
+    
+    SDLSendHapticData* hapticRPC = [[SDLSendHapticData alloc] initWithHapticRectData:hapticRects];
+    [self.connectionManager sendManagerRequest:hapticRPC withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+    }];
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticManager.m]: This should `SDLLogW` the error and response if `response.success` is `NO`

Comment (joeljfischer) [reply SmartDeviceLink/SDLHapticManager.m]: Alternately, it could just pass nil for the handler.

File: SmartDeviceLink/SDLNotificationConstants.h [hunk 19]
```diff
@@ -82,6 +82,7 @@ extern SDLNotificationName const SDLTransportDidConnect;
 extern SDLNotificationName const SDLDidReceiveError;
 extern SDLNotificationName const SDLDidReceiveLockScreenIcon;
 extern SDLNotificationName const SDLDidBecomeReady;
+extern SDLNotificationName const SDLDidProjectionViewUpdate;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLNotificationConstants.h]: I think I would still prefer this to be `SDLDidUpdateProjectionView` as noted in the last review. The style here tends to be `SDL[Did/Will/etc.][verb][object]`

File: SmartDeviceLink/SDLHapticManager.m [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+#pragma mark notifications
+/**
+ Function that gets called when projection view updated notification occurs.
+
+ @param notification object with notification data
+ */
+- (void)sdl_projectionViewUpdated:(NSNotification *)notification {
+    [self updateInterfaceLayout];
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticManager.m]: It does not appear to be tested that a notification send causes an update of the interface layout.

File: SmartDeviceLink/SDLHapticInterface.h [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+@class SDLManager;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol SDLHapticInterface <NSObject>
+
+/**
+ Initializes haptic interface. After initializing the application must call updateInterfaceLayout to process the UIWindow. Application must update later view changes in the window by sending SDLDidProjectionViewUpdate  notification.
+ 
+  @param window UIWindow to initialize
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticInterface.h]: I'd be careful with this; we are merely utilizing the UIWindow, not initializing it.

File: SmartDeviceLink/SDLHapticManager.m [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+#import "SDLTouch.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface SDLHapticManager()
+
+/**
+ The projection window associated with the Haptic Manager
+ */
+@property (nonatomic, strong) UIWindow *projectionWindow;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticManager.m]: Should this be weak because we aren't the primary owner of the window (i.e. is this window expected to be owned by us or by the developer)?

File: SmartDeviceLink/SDLHapticManager.m [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+
+/**
+ Array of focusable view objects extracted from the projection window
+ */
+@property (nonatomic, strong) NSMutableArray<UIView *> *focusableViews;
+
+/**
+ reference to SDLConnectionManager
+ */
+@property (nonatomic, strong) id<SDLConnectionManagerType> connectionManager;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLHapticManager.m]: This should probably be weak?

File: SmartDeviceLink/SDLHapticManager.m (hunk 14 above)
Comment (ChrisB-Elektrobit) [reply SmartDeviceLink/SDLHapticManager.m]: The idea here is that the lowest focusable "leaf" child elements in the view hierarchy should be added as a haptic rects.  The top level focusable view would not be added if it has focusable children.  We added a unit test to cover this case.

File: SmartDeviceLink/SDLHapticManager.m (hunk 15 above)
Comment (ChrisB-Elektrobit) [reply SmartDeviceLink/SDLHapticManager.m]: Correct, for the case where two views are overlapping, the hit test will return `nil`.  This would cover the edge case of overlapping views which could be possible for views on the same hierarchy level.  We added a unit test to cover this as well.

File: SmartDeviceLink/SDLHapticManager.m (hunk 20 above)
Comment (ChrisB-Elektrobit) [reply SmartDeviceLink/SDLHapticManager.m]: We added a test to cover this case as well.

**Comment by joeljfischer** [comment]
LGTM
